### PR TITLE
[ASA-642][RV] add endpoint for ctUtr-crn match

### PR DESCRIPTION
**Pull request by rvijayan**

### CODE REVIEW

File: app/uk/gov/hmrc/agentsubscription/connectors/DesConnector.scala [hunk 1]
```diff
@@ -156,6 +156,16 @@ class DesConnector @Inject() (
     getWithDesHeaders[AgentRecord]("GetAgentRecord", url)
   }
 
+  def getCorporationTaxUtr(crn: Crn)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Utr] = {
+    val encodedCrn = UriEncoding.encodePathSegment(crn.value, "UTF-8")
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentsubscription/connectors/DesConnector.scala]: do we need to encode crn ?  its just a string with no spaces in it ?

File: app/uk/gov/hmrc/agentsubscription/controllers/CTReferenceController.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+@Singleton
+class CTReferenceController @Inject() (service: CTReferenceService)(implicit metrics: Metrics, microserviceAuthConnector: MicroserviceAuthConnector)
+  extends AuthActions(metrics, microserviceAuthConnector) with BaseController {
+
+  def matchCorporationTaxUtr(ctUtr: Utr, crn: Crn): Action[AnyContent] = authorisedWithAgentAffinity { implicit request =>
+    service.matchCorporationTaxUtrWithCrn(ctUtr, crn).map {
+      case Match => Ok
+      case NoMatch | UtrNotFound => NotFound
+      case InvalidCrn => BadRequest
+      case _ => InternalServerError
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentsubscription/controllers/CTReferenceController.scala]: its good to do this validation from the calling service. 
the endpoint in this service just returns the response that it gets from des

Comment (rvijayan) [reply app/uk/gov/hmrc/agentsubscription/controllers/CTReferenceController.scala]: This endpoint matches the utr we supply with the utr in DES `response.` The `case _` is a place holder for unexpected errors(mainly 5xx)  which is not handled in the service.

File: conf/app.routes [hunk 3]
```diff
@@ -6,3 +6,5 @@ POST    /subscription                          @uk.gov.hmrc.agentsubscription.co
 PUT     /subscription                          @uk.gov.hmrc.agentsubscription.controllers.SubscriptionController.updateSubscription()
 
 PUT    /overseas-subscription                 @uk.gov.hmrc.agentsubscription.controllers.SubscriptionController.createOverseasSubscription()
+
+GET    /corporation-tax-utr/:utr/crn/:crn     @uk.gov.hmrc.agentsubscription.controllers.CTReferenceController.matchCorporationTaxUtr(utr:uk.gov.hmrc.agentmtdidentifiers.model.Utr, crn:uk.gov.hmrc.agentsubscription.model.Crn)
```
Comment (sureshhmrc) [inline conf/app.routes]: I mean not to do the utr matching checks in this service, 
we can simply have `GET /corporation-tax/identifiers/crn/AA123456` endpoint and return the des response as is to the caller/frontend , then front-end does the validation checks it requires. 
This way it promotes the re-usability of code, for example: incase other agent services want to re-use the `GET /corporation-tax/identifiers/crn/AA123456` endpoint from this service

Comment (rvijayan) [reply conf/app.routes]: We usually don't create API's which just mirror DES API response. I think the agent-subscription backend should contain the business logic to perform the utr-crn matching operation.  Please have a look at the other implemented API endpoints in agent-subscription as well.

Comment (sureshhmrc) [reply conf/app.routes]: yeah, but unfortunately our mdtp frontends are not just frontends dealing with UI but also business logic.
we can go ahead with your changes , but in that case the use of all these backend endpoints will be limited only to the subscription frontend

Comment (sureshhmrc) [reply conf/app.routes]: let me take another look  at the changes

File: app/uk/gov/hmrc/agentsubscription/service/CTReferenceService.scala [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  case object NoMatch extends CtUtrMatchResponse
+  case object UtrNotFound extends CtUtrMatchResponse
+  case object InvalidCrn extends CtUtrMatchResponse
+  case object UnknownError extends CtUtrMatchResponse
+}
+
+@Singleton
+class CTReferenceService @Inject() (desConnector: DesConnector) {
+
+  def matchCorporationTaxUtrWithCrn(utr: Utr, crn: Crn)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[CtUtrMatchResponse] = {
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentsubscription/service/CTReferenceService.scala]: returning `CtUtrMatchResponse` from this service and dealing those responses in the controller is an overkiil , instead this method can simply return `Future[Boolean]`
method name could be `ctUtrMatchesCrn`  ?

**Pull request merged**